soroban-env-host/src/events/internal.rs

```diff
@@ -130,14 +137,21 @@ pub(crate) struct InternalEventsBuffer {
 impl InternalEventsBuffer {
     // Records an InternalEvent
     pub fn record(&mut self, e: InternalEvent, budget: &Budget) -> Result<(), HostError> {
-        // Metering: we use the cost of instantiating a size=1 `Vec` as an estimate for the cost
-        // `Vec.push(event)`. Because the buffer length may be different on different instances
-        // due to diagnostic events and we need a deterministic cost across all instances,
-        // the cost needs to be amortized and buffer size-independent.
-
-        if let InternalEvent::Contract(_) = e {
-            Vec::<(InternalEvent, EventError)>::charge_bulk_init_cpy(1, budget)?;
+        // Metering: we use the cost of instantiating a size=1 `Vec` as an estimate
+        // for the cost `Vec.push(event)`. Because the buffer length may be different
+        // on different instances due to diagnostic events and we need a deterministic
+        // cost across all instances, the cost needs to be amortized and buffer
+        // size-independent.
+        let charge_internal_event_push =
+            || Vec::<(InternalEvent, EventError)>::charge_bulk_init_cpy(1, budget);
+
+        match e {
+            InternalEvent::Contract(_) => charge_internal_event_push()?,
+            InternalEvent::Diagnostic(_) => {
+                budget.with_internal_mode(charge_internal_event_push, || ())
+            }
         }
+
         self.vec.push((e, EventError::FromSuccessfulCall));
```

Comment:
I'm not 100% sure this is right. I think we want to guard pushing the Diagnostic event on debug mode too, right? Like don't even record it if we're not in debug mode (usually this is filtered out by a caller I think but this function is marked `pub` so .. theoretically someone could call without). And also if the internal / debug charge fails, I think we should avoid pushing the corresponding diagnostic event.

Reply:
This is fixed.